### PR TITLE
[FEATURES] Predict with Particular Model

### PR DESCRIPTION
**Pull request by MatthewK3023**
Signed-off-by: Kenyo <kenyo3023@gmail.com>

### CODE REVIEW

File: src/decanter/core/core_api/predict_input.py [hunk 1]
```diff
@@ -27,7 +28,7 @@ class PredictInput:
 
     """
     def __init__(
-            self, data, experiment, callback=None,
+            self, data, experiment, select_by='by_CVbest', select_model_by='mse', select_model_id='', callback=None,
```
Comment (NoctisHong) [inline src/decanter/core/core_api/predict_input.py:31]: select_model_id=None

File: src/decanter/core/core_api/predict_input.py [hunk 2]
```diff
@@ -49,15 +50,43 @@ def __init__(
...
 
         Returns:
             :obj:`dict`
         """
-        setattr(self.pred_body, 'data_id', self.data.id)
-        setattr(self.pred_body, 'model_id', self.experiment.best_model.id)
+
+        # '''
+        if self.select_by == 'by_CVbest':
+            print(self.select_by, self.select_model_by, '!!!!!!!!!!!')
```
Comment (NoctisHong) [inline src/decanter/core/core_api/predict_input.py:68]: use logger instead of print

File: src/decanter/core/jobs/experiment.py [hunk 3]
```diff
@@ -94,6 +94,9 @@ def create(cls, exp_id, name=None):
     def update_result(self, task_result):
         """Update Job's attribute from Task's result."""
         self.get_best_model()
+        self.recommend_model_id = {rec_dict['evaluator']: rec_dict['model_id'] \
```
Comment (CliffLin) [inline src/decanter/core/jobs/experiment.py:97]: I guess that it's my mistake.. this data has been store at 'recommendations ' and `models`. Those data might be updated in task.CoreTask.update_task_response

File: src/decanter/core/core_api/predict_input.py [hunk 4]
```diff
@@ -49,15 +50,43 @@ def __init__(
...
+        # else:
+        elif self.select_by == 'by_RC':
+            print(self.select_by, self.select_model_by)
+            if self.select_model_by in self.experiment.recommend_model_id.keys():
+                setattr(self.pred_body, 'data_id', self.data.id)
+                setattr(self.pred_body, 'model_id', self.experiment.recommend_model_id[self.select_model_by])
+            else:
+                logger.error('[%s] no such Metric on model', class_, self.select_model_by) 
+        elif self.select_by == 'by_index':
+            print(self.select_by, self.select_model_by)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:80]: remove used code

File: src/decanter/core/core_api/predict_input.py [hunk 5]
```diff
@@ -49,15 +50,43 @@ def __init__(
...
-        setattr(self.pred_body, 'model_id', self.experiment.best_model.id)
+
+        # '''
+        if self.select_by == 'by_CVbest':
+            print(self.select_by, self.select_model_by, '!!!!!!!!!!!')
+            setattr(self.pred_body, 'data_id', self.data.id)
+            setattr(self.pred_body, 'model_id', self.experiment.best_model.id)
+        # else:
+        elif self.select_by == 'by_RC':
+            print(self.select_by, self.select_model_by)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:73]: ditto

File: src/decanter/core/core_api/predict_input.py [hunk 6]
```diff
@@ -49,15 +50,43 @@ def __init__(
         self.pred_body = CoreBody.PredictBody.create(
             data_id='tmp_data_id', model_id='tmp_model_id', callback=callback,
             keep_columns=keep_columns, threshold=threshold, version=version)
+        # '''
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:53]: ditto

**Comment by MatthewK3023** [comment]
I sent a new pull request, named [FEATURES] Remove Unused Code 
截圖 2020-08-15 上午1 47 55
this part, i will check and fix it on Monday, because It seems impossible to connect to the server without being in the company on weekends

**Comment by CliffLin** [comment]
Please leave the comment in the thread, or just click "Resolve" button if you commit

File: .gitignore [hunk 7]
```diff
@@ -141,3 +141,7 @@ dmypy.json
 # vs code
 .vscode
 
+# testing files
+test-decanter.ipynb
```
Comment (CliffLin) [inline .gitignore:145]: I think those 3 files are your test file. These files won't apprear in developers' env.

File: src/decanter/core/core_api/predict_input.py [hunk 8]
```diff
@@ -6,6 +6,8 @@
 
 from decanter.core.core_api import CoreBody
 
+import logging
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:9]: import grouping. Move it to line 5

File: src/decanter/core/core_api/predict_input.py [hunk 9]
```diff
@@ -49,21 +51,39 @@ def __init__(
...
         """Using pred_body to create the JSON request body for prediction.
 
         Returns:
             :obj:`dict`
         """
-        setattr(self.pred_body, 'data_id', self.data.id)
-        setattr(self.pred_body, 'model_id', self.experiment.best_model.id)
+        class_ = self.__class__.__name__
+        if self.select_by == 'by_CVbest':
+            setattr(self.pred_body, 'data_id', self.data.id)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:66]: this line is recurring.

File: src/decanter/core/core_api/predict_input.py [hunk 10]
```diff
@@ -49,21 +51,39 @@ def __init__(
...
+            for rec in self.experiment.recommendations:
+                if self.select_model_by == rec['evaluator']:
+                    setattr(self.pred_body, 'data_id', self.data.id)
+                    setattr(self.pred_body, 'model_id', rec['model_id'])
+        elif self.select_by == 'by_model_id':
+            if self.select_model_id in self.experiment.models:
+                setattr(self.pred_body, 'data_id', self.data.id)
+                setattr(self.pred_body, 'model_id', self.select_model_id)
+            else:
+                logger.error('[%s] no such model ID', class_, self.select_model_id)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:78]: if user pass wrong model id, how will it predict?

File: src/decanter/core/jobs/experiment.py (hunk 3 above)
Comment (MatthewK3023) [reply src/decanter/core/jobs/experiment.py]: fix, i use experiment.recommendations to replace the self.recommend_model_id

File: src/decanter/core/core_api/predict_input.py (hunk 5 above)
Comment (MatthewK3023) [reply src/decanter/core/core_api/predict_input.py]: fix

File: src/decanter/core/core_api/predict_input.py (hunk 10 above)
Comment (MatthewK3023) [reply src/decanter/core/core_api/predict_input.py]: if user pass wrong model id or metric not in the model metric list, will raise the error.
and the [data_id and model_id] will become ['tmp_data_id' and 'tmp_model_id']

File: src/decanter/core/core_api/predict_input.py (hunk 9 above)
Comment (MatthewK3023) [reply src/decanter/core/core_api/predict_input.py]: sorry, I may not understand the meaning of recursion here,  This recursive situation still appears in the new PR?

File: src/decanter/core/core_api/predict_input.py (hunk 8 above)
Comment (MatthewK3023) [reply src/decanter/core/core_api/predict_input.py]: fix

File: src/decanter/core/core_api/predict_input.py [hunk 11]
```diff
@@ -27,7 +28,7 @@ class PredictInput:
 
     """
     def __init__(
-            self, data, experiment, callback=None,
+            self, data, experiment, select_model='best', callback=None,
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:31]: I thought this parameter is weird. How about make `predict_method` and `predict_opt` as parameters?
e.g. predict_method = predict_method.CV_BEST, predict_opt = None or predict_method = predict_method.MODEL_ID, predict_opt = 00aabbcc11

File: src/decanter/core/core_api/predict_input.py (hunk 10 above)
Comment (CliffLin) [reply src/decanter/core/core_api/predict_input.py]: where do you throw the error?

File: src/decanter/core/core_api/predict_input.py (hunk 11 above)
Comment (CliffLin) [reply src/decanter/core/core_api/predict_input.py:31]: After updated the parameters, please update the docstring.

File: src/decanter/core/core_api/predict_input.py [hunk 12]
```diff
@@ -49,21 +61,36 @@ def __init__(
...
         Returns:
             :obj:`dict`
         """
+        if self.select_model == 'best':
+            select_model_id = self.experiment.best_model.id
+        elif self.select_model == 'model_id':
+            if self.select_opt in self.experiment.models:
+                select_model_id = self.select_opt
+            else:
+                logger.error('[%s] Invalid input model ID', self.__class__.__name__, self.select_opt)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:79]: should we raise a exception over here?

File: src/decanter/core/core_api/predict_input.py [hunk 13]
```diff
@@ -49,21 +61,36 @@ def __init__(
...
+        elif self.select_model == 'recommendation':
+            for rec in self.experiment.recommendations:
+                if self.select_opt == rec['evaluator']:
+                   select_model_id = rec['model_id']
+            if 'select_model_id' not in locals().keys():
+                logger.error('[%s] Invalid input metric', self.__class__.__name__, self.select_opt)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:85]: ditto

**Comment by CliffLin** [comment]
https://mobagel.github.io/decanter-ai-core-sdk/notes/contributing.html#code-contributions
Please follow the steps to make sure unittest and lint are passed.

**Review by CliffLin — COMMENTED** [review]
Please run `make lint test` before commit

**Review by CliffLin — APPROVED** [review]

**Review by CliffLin — COMMENTED** [review]
https://mobagel.github.io/decanter-ai-core-sdk/notes/contributing.html#code-contributions

**Review by schwannden — APPROVED** [review]

File: src/decanter/core/core_api/predict_input.py [hunk 14]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
...
+            select_model (:obj: `str`, optional): Methods of screening models
+                `best`: based on best CV
+                `model_id`: based on given Model ID
+                `recommendation`: recommended model of the given metric
+                Defaults to 'best'.
+            select_opt (:obj: `str`, optional): Based on the options required by 
+                the select_model.
+                value with select_model case:
+                    `best`          : None
+                    `model_id`      : given the model ID, the ObjectId
```
Comment (schwannden) [inline src/decanter/core/core_api/predict_input.py:49]: given the model ID (model ID will be in the format of ObjectID)

File: src/decanter/core/core_api/predict_input.py [hunk 15]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
...
+                keep_columns=None, threshold=None, version=None):
         """
         Init Predict Input
 
         Args:
             data (:class:`~decanter.core.jobs.data_upload.DataUpload`): Test data.
             experiment (:class:`~decanter.core.jobs.experiment.Experiment`):
                 Experiment from training.
+            select_model (:obj: `str`, optional): Methods of screening models
+                `best`: based on best CV
```
Comment (schwannden) [inline src/decanter/core/core_api/predict_input.py:41]: predict with the best model scored on cv average

File: src/decanter/core/core_api/predict_input.py [hunk 16]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
...
+                `model_id`: based on given Model ID
```
Comment (schwannden) [inline src/decanter/core/core_api/predict_input.py:42]: predict with the model designated by model_id

File: src/decanter/core/core_api/predict_input.py [hunk 17]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
...
+                `recommendation`: recommended model of the given metric
```
Comment (schwannden) [inline src/decanter/core/core_api/predict_input.py:43]: predict with the recommended model

**Review by schwannden — CHANGES_REQUESTED** [review]
please make document more detailed and readable

File: src/decanter/core/core_api/predict_input.py [hunk 18]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
 
     """
     def __init__(
-            self, data, experiment, callback=None,
-            keep_columns=None, threshold=None, version=None):
+            self, data, experiment, select_model='best', select_opt=None, callback=None,
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:31]: suffix space

File: src/decanter/core/core_api/predict_input.py [hunk 19]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
 
     """
     def __init__(
-            self, data, experiment, callback=None,
-            keep_columns=None, threshold=None, version=None):
+            self, data, experiment, select_model='best', select_opt=None, callback=None, 
+                keep_columns=None, threshold=None, version=None):
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:32]: prefix space

File: src/decanter/core/core_api/predict_input.py (hunk 16 above)
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:42]: "m"odel

File: src/decanter/core/core_api/predict_input.py [hunk 20]
```diff
@@ -27,15 +28,26 @@ class PredictInput:
...
                 Experiment from training.
+            select_model (:obj: `str`, optional): Methods of screening models
+                `best`: based on best CV
+                `model_id`: based on given Model ID
+                `recommendation`: recommended model of the given metric
+                Defaults to 'best'.
+            select_opt (:obj: `str`, optional): Based on the options required by 
+                the select_model.
+                value with select_model case:
+                    `best`          : None
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:48]: `best`          : None -> `best` : None

File: src/decanter/core/core_api/predict_input.py (hunk 14 above)
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:49]: ditto

File: src/decanter/core/core_api/predict_input.py [hunk 21]
```diff
@@ -49,21 +61,38 @@ def __init__(
...
+                select_model_id = self.select_opt
+            else: 
+                logger.error('[%s] Invalid input model ID', self.__class__.__name__, self.select_opt)
+                raise ValueError('Invalid input model ID: %s'%self.select_opt)
+        elif self.select_model == 'recommendation':
+            for rec in self.experiment.recommendations:
+                if self.select_opt == rec['evaluator']:
+                   select_model_id = rec['model_id']
+            if 'select_model_id' not in locals().keys():
+                logger.error('[%s] Invalid input metric', self.__class__.__name__, self.select_opt)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:86]: does it work? There is just one `%s`, but you put 2 variables into it.

File: src/decanter/core/core_api/predict_input.py [hunk 22]
```diff
@@ -49,21 +61,38 @@ def __init__(
...
+            else: 
+                logger.error('[%s] Invalid input model ID', self.__class__.__name__, self.select_opt)
+                raise ValueError('Invalid input model ID: %s'%self.select_opt)
+        elif self.select_model == 'recommendation':
+            for rec in self.experiment.recommendations:
+                if self.select_opt == rec['evaluator']:
+                   select_model_id = rec['model_id']
+            if 'select_model_id' not in locals().keys():
+                logger.error('[%s] Invalid input metric', self.__class__.__name__, self.select_opt)    
+                raise ValueError('Invalid input metric: %s'%self.select_opt)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:87]: ditto

File: src/decanter/core/core_api/predict_input.py [hunk 23]
```diff
@@ -49,21 +61,38 @@ def __init__(
...
             :obj:`dict`
         """
+        if self.select_model == 'best':
+            select_model_id = self.experiment.best_model.id
+        elif self.select_model == 'model_id':
+            if self.select_opt in self.experiment.models:
+                select_model_id = self.select_opt
+            else: 
+                logger.error('[%s] Invalid input model ID', self.__class__.__name__, self.select_opt)
+                raise ValueError('Invalid input model ID: %s'%self.select_opt)
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:80]: add space around `%`

File: src/decanter/core/core_api/predict_input.py [hunk 24]
```diff
@@ -3,9 +3,10 @@
 # pylint: disable=too-few-public-methods
 """Settings for the PredictResult and PredictTSResult"""
 import json
-
+import logging
 from decanter.core.core_api import CoreBody
```
Comment (CliffLin) [inline src/decanter/core/core_api/predict_input.py:7]: add new line between our lib and standard lib

File: src/decanter/core/core_api/predict_input.py (hunk 16 above)
Comment (CliffLin) [reply src/decanter/core/core_api/predict_input.py:42]: I think schwannden's comment might be better

**Review by schwannden — APPROVED** [review]

**Review by NoctisHong — APPROVED** [review]

**Review by NoctisHong — APPROVED** [review]

**Pull request merged**